RecoMuon/MuonIdentification/plugins/cuts/MuonPOGStandardCuts.cc

```diff
@@ -0,0 +1,69 @@
+#include "PhysicsTools/SelectorUtils/interface/CutApplicatorBase.h"
+#include "PhysicsTools/SelectorUtils/interface/CutApplicatorWithEventContentBase.h"
+#include "DataFormats/MuonReco/interface/Muon.h"
+#include "DataFormats/MuonReco/interface/MuonSelectors.h"
+
+class MuonPOGStandardCut : public CutApplicatorWithEventContentBase
+{
+public:
+  MuonPOGStandardCut(const edm::ParameterSet& c);
+
+  result_type operator()(const reco::MuonPtr&) const override final;
+  CandidateType candidateType() const override final { return MUON; }
+
+  void setConsumes(edm::ConsumesCollector&) override final;
+  void getEventContent(const edm::EventBase&) override final;
+
+private:
+  enum CutType {
+    LOOSE, MEDIUM, TIGHT, SOFT, HIGHPT,
+    NONE
+  } cutType_;
+  edm::Handle<reco::VertexCollection> vtxs_;
+};
+DEFINE_EDM_PLUGIN(CutApplicatorFactory,
+                  MuonPOGStandardCut, "MuonPOGStandardCut");
+
+// Define constructors and initialization routines
+MuonPOGStandardCut::MuonPOGStandardCut(const edm::ParameterSet& c):
+  CutApplicatorWithEventContentBase(c)
+{
+  const auto cutTypeName = c.getParameter<std::string>("idName");
+  if ( cutTypeName == "loose") cutType_ = LOOSE;
+  else if ( cutTypeName == "tight") cutType_ = TIGHT;
+  else if ( cutTypeName == "medium") cutType_ = MEDIUM;
+  else if ( cutTypeName == "soft" ) cutType_ = SOFT;
+  else if ( cutTypeName == "highpt" ) cutType_ = HIGHPT;
+  else
+  {
+    edm::LogError("MuonPOGStandardCut") << "Wrong cut id name, " << cutTypeName;
+    cutType_ = NONE;
+  }
+
+  contentTags_.emplace("vertices", c.getParameter<edm::InputTag>("vertexSrc"));
+}
+
+void MuonPOGStandardCut::setConsumes(edm::ConsumesCollector& cc)
+{
+  auto vtcs = cc.consumes<reco::VertexCollection>(contentTags_["vertices"]);
+  contentTokens_.emplace("vertices", vtcs);
+}
+
+void MuonPOGStandardCut::getEventContent(const edm::EventBase& ev)
+{
+  ev.getByLabel(contentTags_["vertices"], vtxs_);
+}
+
+// Functors for evaluation
+CutApplicatorBase::result_type MuonPOGStandardCut::operator()(const reco::MuonPtr& cand) const
+{
+  if ( cutType_ == LOOSE ) return muon::isLooseMuon(*cand);
+  else if ( cutType_ == TIGHT ) return muon::isTightMuon(*cand, vtxs_->at(0));
```

Comment:
Since there would be no other way to calculate the ID variable being cut
upon, I believe an exception from the missing is the best option.

Best,
-L

On Mon, Jun 1, 2015 at 7:11 PM, Carl Vuosalo <notifications@github.com>
wrote:

> In RecoMuon/MuonIdentification/plugins/cuts/MuonPOGStandardCuts.cc
> <https://github.com/cms-sw/cmssw/pull/9273#discussion_r31446898>:
>
> > +void MuonPOGStandardCut::setConsumes(edm::ConsumesCollector& cc)
> > +{
> > +  auto vtcs = cc.consumes<reco::VertexCollection>(contentTags_["vertices"]);
> > +  contentTokens_.emplace("vertices", vtcs);
> > +}
> > +
> > +void MuonPOGStandardCut::getEventContent(const edm::EventBase& ev)
> > +{
> > +  ev.getByLabel(contentTags_["vertices"], vtxs_);
> > +}
> > +
> > +// Functors for evaluation
> > +CutApplicatorBase::result_type MuonPOGStandardCut::operator()(const reco::MuonPtr& cand) const
> > +{
> > +  if ( cutType_ == LOOSE ) return muon::isLooseMuon(*cand);
> > +  else if ( cutType_ == TIGHT ) return muon::isTightMuon(*cand, vtxs_->at(0));
>
> OK. I can only guess at the possible conditions that would trigger this
> problem. Maybe data corruption or a coding error where getEventContent()
> was not called properly could cause this exception. It seems a bit extreme
> to me to throw an exception just because of a difficulty in classifying a
> muon for ID. Is an exception the most elegant way to recover from this
> problem?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9273/files#r31446898>.
>